server/storage/backend/batch_tx_test.go

```diff
@@ -205,3 +207,62 @@ func TestBatchTxBatchLimitCommit(t *testing.T) {
 		return nil
 	})
 }
+
+func TestRangeAfterDeleteMatch(t *testing.T) {
+	b, _ := betesting.NewTmpBackend(t, time.Hour, 10000)
+	defer betesting.Close(t, b)
+
+	tx := b.BatchTx()
+
+	tx.Lock()
+	tx.UnsafeCreateBucket(schema.Test)
+	tx.UnsafePut(schema.Test, []byte("foo"), []byte("bar"))
+	tx.Unlock()
+	tx.Commit()
+
+	checkTxnResponseMatch(t, b.BatchTx(), b.ReadTx(), []byte("foo"), nil, 0)
+
+	tx.Lock()
+	tx.UnsafeDelete(schema.Test, []byte("foo"))
+	tx.Unlock()
+
+	checkTxnResponseMatch(t, b.BatchTx(), b.ReadTx(), []byte("foo"), nil, 0)
+}
+
+func TestRangeAfterCommitMatch(t *testing.T) {
+	b, _ := betesting.NewTmpBackend(t, time.Hour, 10000)
+	defer betesting.Close(t, b)
+
+	tx := b.BatchTx()
+
+	tx.Lock()
+	tx.UnsafeCreateBucket(schema.Test)
+	tx.Unlock()
+
+	tx.Lock()
+	tx.UnsafePut(schema.Test, []byte("foo"), []byte("bar"))
+	tx.Unlock()
+	tx.Commit()
+
+	checkTxnResponseMatch(t, b.BatchTx(), b.ReadTx(), []byte("foo"), []byte("foo3"), 1)
+
+	tx.Lock()
+	tx.UnsafePut(schema.Test, []byte("foo2"), []byte("bar2"))
+	tx.Unlock()
+
+	checkTxnResponseMatch(t, b.BatchTx(), b.ReadTx(), []byte("foo"), []byte("foo3"), 1)
```

Comment:
The main problem is it doesn't guarantee the order.  
- `batchTx.UnsafeRange` reads the bbolt directly, it's the correct results. 
- While `readTx.UnsafeRange` reads the buffer firstly, and then read from bbolt if the results returned from buffer doesn't meet the limit. Otherwise, if the data returned from buffer meets the limit, it will not continue to read from bbolt. But the results returned from buffer may not be correctly ordered against the data in bbolt.

In this case, if you inserts foo2 firstly, then foo, then the case will pass. 

My immediate thought on the correct solution should be:
- if `endKey == nil` or `limit == math.MaxInt64`, the existing logic is correct.
- Otherwise, we should read at most `limit` records from both buffer and bbolt respectively, merge the results and sort, finally return the correct `limit` number of records.

Reply:
Right, what you proposed makes sense. What worries me is that this was totally unknown behavior which means huge gaps in our testing.

Reply:
It might not be an issue. 

We do not have to guarantee the order here. If the client cares about the order of the result, then etcd will set the limit as 0, which means no limit.

https://github.com/etcd-io/etcd/blob/4e986363a371ad88479cb8f9b248d16ea83abc87/server/etcdserver/txn/txn.go#L153-L159

There are two special internal use cases. The range [key, endKey] must be only in bbolt for both cases. 
The first one is compaction.
https://github.com/etcd-io/etcd/blob/4e986363a371ad88479cb8f9b248d16ea83abc87/server/storage/mvcc/kvstore_compaction.go#L53

The second use case is recovering,
https://github.com/etcd-io/etcd/blob/4e986363a371ad88479cb8f9b248d16ea83abc87/server/storage/mvcc/kvstore.go#L341

Reply:
I don't think this is an issue for clients. `tx.UnsafeRange` with range limit is mainly used internally. For the client Range requests, it maps keys to revs through the keyIndex first
```
revpairs, total := tr.s.kvindex.Revisions(key, end, rev, int(ro.Limit))
```
then each rev key is queried individually 
```
for i, revpair := range revpairs[:len(kvs)] {
...
		revBytes = RevToBytes(revpair, revBytes)
                // UnsafeRange is called with nil endkey.
		_, vs := tr.tx.UnsafeRange(schema.Key, revBytes, nil, 0)
...
```
So the client range results are always ordered due to the btree in kvindex.

Reply:
Just to make it clear, I agree it might not be an issue, however based on other discovery (deletion not being passed to read cache) I think we should double check. I think it's an unobvious behavior that:
* should be properly documented
* we should double check if there is any internal misuse
* we could consider reflecting in in function names.

Reply:
Please raise an issue to track it.